879-Admin-Comment-View

jacknussbaum opened the pull request:
Resolves #879 

Allows users to view the Google Street View location where a comment was made. (Accessed through the comments table on the admin interface)


![image](https://user-images.githubusercontent.com/71039842/97484891-97b30980-1916-11eb-889c-b8474745e2d0.png)

jacknussbaum commented:
Should be all updated! Now includes labels in the comment view windows.

Labels:
![image](https://user-images.githubusercontent.com/71039842/99726626-8cfa1900-2a6b-11eb-865b-78e8f4e9cd25.png)
![image](https://user-images.githubusercontent.com/71039842/99726714-b155f580-2a6b-11eb-9ff3-13d5ae705010.png)

misaugstad commented on app/models/audit/AuditTaskCommentTable.scala:
```diff
@@ -12,7 +12,7 @@ case class AuditTaskComment(auditTaskCommentId: Int, auditTaskId: Int, missionId
                             pitch: Option[Double], zoom: Option[Int], lat: Option[Double], lng: Option[Double],
                             timestamp: Timestamp, comment: String)
-case class GenericComment(commentType: String, username: String, gsvPanoramaId: Option[String], timestamp: Timestamp, comment: String)
+case class GenericComment(commentType: String, username: String, gsvPanoramaId: Option[String], timestamp: Timestamp, comment: String, heading: Option[Double], pitch: Option[Double], zoom: Option[Int], labelId: Int)
```
I think what you want is to make the `labelId` an option type, and then instead of `0` for the audit task comments, you would use `None`. Then in the JavaScript, you would be checking of the `labelId` is `null` instead of 0.

jacknussbaum replied:
I have the code like this and I keep getting the error shown.. any thoughts?
![image](https://user-images.githubusercontent.com/71039842/100801738-3f6f9b80-33dd-11eb-9e6d-b3d2345a785f.png)

![image](https://user-images.githubusercontent.com/71039842/100801693-3088e900-33dd-11eb-8d5e-fb250e8d774c.png)

![image](https://user-images.githubusercontent.com/71039842/100801811-59a97980-33dd-11eb-83e6-6866e0b8caa2.png)

misaugstad replied:
Try replacing `None` with `(None: Option[Int])` in the query.

misaugstad commented on app/views/admin/index.scala.html:
```diff
@@ -322,7 +322,7 @@ <h1>Comments</h1>
                                     </td>
                                     <td class = 'timestamp'>@comment.timestamp</td>
-                                    <td>@comment.gsvPanoramaId</td>
+                                    <td><a class="showCommentLocation" href="#" data-heading="@comment.heading" data-pitch="@comment.pitch" data-zoom="@comment.zoom" data-label-Id="@comment.labelId">@comment.gsvPanoramaId</a></td>
```
Can you spread this out more so that the line is less than 120 characters?

misaugstad commented on public/javascripts/Admin/src/Admin.GSVCommentView.js:
```diff
@@ -0,0 +1,68 @@
+
+    var _init = function() {
+        self.panoProp = new PanoProperties();
+
```
I don't think the extra line needs to be here. The function is only 2 lines long, separating out the different sections with the extra line seems like overkill :wink:

misaugstad commented on public/javascripts/Admin/src/Admin.GSVCommentView.js:
```diff
@@ -0,0 +1,68 @@
+
+    function _resetModal() {
+        var modalText =
+            '<div class="modal fade" id="labelModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel">'+
```
I realize that this is done wrong in other places, but can you make add a space between the `'` and the `+` on each line?

misaugstad commented on public/javascripts/Admin/src/Admin.GSVCommentView.js:
```diff
@@ -0,0 +1,68 @@
+
+    function _resetModal() {
+        var modalText =
+            '<div class="modal fade" id="labelModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel">'+
```
Can you actually also modify this HTML (and the same code in `GSVLabelView`) to user `this-case` instead of `camelCase` for element IDs like `labelModal` -> `label-modal`

jacknussbaum replied:
Have you heard that referred to as "kebab-case"? Fixed those instances

misaugstad replied:
I had not! Thank you :)

misaugstad commented on public/javascripts/Admin/src/Admin.GSVCommentView.js:
```diff
@@ -0,0 +1,68 @@
+                        '<div class="modal-body">'+
+                            '<div id="svholder" style="width: 810px; height:540px">'+
+                        '</div>'
+                        +
```
This `+` should probably be on the previous line

misaugstad commented on public/javascripts/Admin/src/Admin.GSVCommentView.js:
```diff
@@ -0,0 +1,68 @@
+                        '</div>'
+                        +
+                        '<div id="button-holder">' + '</div>' +
+                '</div>'+
```
Can you make these aligned with where they normally would be like the rest of the code here?

misaugstad commented on public/javascripts/Admin/src/Admin.GSVCommentView.js:
```diff
@@ -0,0 +1,68 @@
+            labelMetadata['canvas_width'], labelMetadata['canvas_height'], labelMetadata['heading'],
+            labelMetadata['pitch'], labelMetadata['zoom']);
+        self.panorama.setLabel(adminPanoramaLabel);
+
```
Definitely don't need this blank line here

misaugstad commented on public/javascripts/Admin/src/Admin.GSVCommentView.js:
```diff
@@ -0,0 +1,68 @@
+    self.showCommentGSV = showCommentGSV;
+
+    return self;
+}
```
Main comment after looking at this file is that you are way too generous with blank lines lol. They can be helpful to separate out sections of code and make things more readable. But I think you've gone overboard here, and it makes it so that you can see less code at one time. I didn't comment about every time I thought it was overkill, I would just suggest dialing it back a bit :)

jacknussbaum replied:
You're very right. Cleaned those up haha

misaugstad reviewed (changes requested):
Mostly just some small changes to make! Great work on this though :grin:

misaugstad commented:
@jacknussbaum is this ready for review again or is there still more work to do?

jacknussbaum commented:
Fixed all the requests, ready for review again!

misaugstad commented on app/views/admin/index.scala.html:
```diff
@@ -322,7 +322,12 @@ <h1>Comments</h1>
                                     <td class = 'timestamp'>@comment.timestamp</td>
-                                    <td>@comment.gsvPanoramaId</td>
+                                    <td>
+                                        <a class="showCommentLocation" href="#" 
```
Change from camelCase to kebab-case

jacknussbaum replied:
There are some other instances throughout this file of camelCase ids from older code, should I change those as well?

misaugstad commented on app/views/admin/index.scala.html:
```diff
@@ -322,7 +322,12 @@ <h1>Comments</h1>
                                     <td class = 'timestamp'>@comment.timestamp</td>
-                                    <td>@comment.gsvPanoramaId</td>
+                                    <td>
+                                        <a class="showCommentLocation" href="#" 
```
And I feel like the entire `<a>` tag should probably be on one line... Looking for
```
<a blah blah blah>
    @comment.gsvPanoramaId
</a>
```

jacknussbaum replied:
This was the line that you requested be shortened to under 120 characters. I can put the whole tag back on one line, but it would exceed that.

misaugstad replied:
hmm, okay well right now it is still over 120 characters, looks like the 2nd line is at 160 (you have to include the spaces). and I'm seeing that I didn't fully describe what I meant. In the HTML, I think that it is okay-ish to have lines over 120 characters; it's kind of inevitable. BUT if you have a really long line, then I think that by default you shouldn't be stacking multiple tags to make it even longer. So the issue I was really getting at was that you had
```
<td><a blah blah>@comment.gsvPanoramaId</a></td>
```

and I think it's better to split into multiple lines if it goes over 120 characters (including spaces). So you want to spread the `td` and `a` and their closing tags:
```
<td>
    <a blah blah>
        stuff in the a tag
    </a>
</td>
```

Even if one of those lines becomes too long, I think this makes the most sense visually.

jacknussbaum replied:
Ohhh yes for sure. Sorry for the misunderstanding, that makes total sense.

misaugstad reviewed (changes requested):
I added some super super minor style comments. BUT it just totally isn't working anymore :grin: It looks like I'm getting a different error for validation comments vs audit comments too lol.

jacknussbaum commented:
What type of errors are you getting?

misaugstad commented:
>What type of errors are you getting?

Is it still working for you still?

jacknussbaum commented:
Hmm I am getting the message that the imagery is no longer available for those comment locations.. is this what you were getting? I had no errors when I pushed the initial PR for review, there were some files updated since then throughout the codebase...

Is that the issue you were running into?

misaugstad commented:
I believe that is the issue I was running into for comments on the validation page. And for comments from the audit page, I was seeing a different error in the console about the /adminapi/label/id endpoint. I did only test that part with one label though.

Even if it was related to recent changes in the codebase, can you investigate?

jacknussbaum commented:
Yes I will get to the bottom of this!

jacknussbaum commented:
Went back in time... checked out a commit from back when it was working properly. See below:
![image](https://user-images.githubusercontent.com/71039842/101588409-32354c80-399b-11eb-8080-3ad98ef1cde7.png)
 The weird thing is that now I am occasionally getting the 'image not available' error even in the old version. I am getting it for panos that I am 100% positive I was able to load previously. 
![image](https://user-images.githubusercontent.com/71039842/101588495-690b6280-399b-11eb-822c-a465a31ccbe2.png)

misaugstad commented:
>I know I tested all the comments on the first page, without any errors before I submitted the initial PR. This is an error from one of those comments^^^^

Note that Google can remove imagery whenever, so it is entirely possible that the imagery is just no longer there. You can check if there is imagery in a location by directly using a URL like [http://maps.google.com/cbk?output=tile&zoom=3&x=2&y=2&cb_client=maps_sv&fover=2&onerr=3&renderer=spherical&v=4&panoid=MoKLKl36pQ5XeQQAR7x__Q](http://maps.google.com/cbk?output=tile&zoom=3&x=2&y=2&cb_client=maps_sv&fover=2&onerr=3&renderer=spherical&v=4&panoid=MoKLKl36pQ5XeQQAR7x__Q). Just replace the pano ID at the end with the appropriate pano id to check.

misaugstad commented:
Can I ask you to go through and figure what commit (or merge commit) was the one that broke this? Find an image that you know works in an old version, and keep going back to check that image to find the exact commit that causes the issue. Then we can both take a look at the changelog for that commit to see what might be causing the issue.

jacknussbaum commented:
Commit: 02e0db4489548d92f3c99734f6f53c738c88d2d7 is the last commit where it was functioning properly. This is the commit titled: Merge branch 'develop' into 879-Comment-GSV-View. Looked at some of the changes made by the commit that directly follows it (the first instance where it no longer works) and I reverted some of them and could not get it to work....

THEN I changed some indentation... and it worked?!

Full functionality:
![image](https://user-images.githubusercontent.com/71039842/104685116-2da53780-56af-11eb-8c86-b58e65a55037.png)

NO functionality:
![image](https://user-images.githubusercontent.com/71039842/104685189-4f9eba00-56af-11eb-9985-df749bc3d042.png)

All I did is make the <a> tag indented to span multiple lines so it didn't form one long line, and that broke it. Should this be expected? This goes against all my understanding of how indentation, compiling etc. works.

misaugstad commented:
Yeah wow I have nooo idea why that would happen. Like the template scala can have some weird syntax quirks, but that one seems unfamiliar to me.

jacknussbaum commented:
This PR is once again functioning. Required some serious hours of bug tracing and could not be happier that it is once again working. Should be ready for review
![image](https://user-images.githubusercontent.com/71039842/105227259-efc46b00-5b15-11eb-8b81-750051646f39.png)

misaugstad commented on app/views/admin/index.scala.html:
```diff
@@ -322,7 +322,10 @@ <h1>Comments</h1>
-                                    <td>@comment.gsvPanoramaId</td>
+                                    <td>
+                                        <a class="show-comment-location" href="#" 
+                                        data-heading="@comment.heading" data-pitch="@comment.pitch" data-zoom="@comment.zoom" data-label-id="@comment.labelId">@comment.gsvPanoramaId</a>
```
I fixed the `data-label-Id` attribute ID capitalization to be `data-label-id`

misaugstad commented on public/javascripts/Admin/src/Admin.GSVCommentView.js:
```diff
@@ -0,0 +1,58 @@
+        });
+        self.panorama.setPano(panoId, heading, pitch, zoom);
+        
+        if(labelId) {
```
I changed this from `if (labelId != null)` to `if (labelId)`. Because I think that the label Id was an empty string, not null.

misaugstad commented on public/javascripts/Admin/src/Admin.js:
```diff
@@ -69,6 +69,21 @@ function Admin(_, $, difficultRegionIds) {
+    function initializeAdminGSVCommentWindow(){
+        $('.show-comment-location').click(function(e) { 
+            e.preventDefault();
+            var heading = parseFloat($(this).data('heading'));
```
I was running into a bug frequently where the GSV imagery wouldn't show up even when there was imagery for it. I was seeing an error that said something like "`a.heading.toFixed` is not a function". I eventually figured out that it was because when you're reading the data using `$(this).data('heading')`, it usually comes back as a string. So I just parsed all these values to the appropriate data type, and it's all working now :ok_hand:

misaugstad reviewed (approved):
I made some changes myself just so we could finish up this PR. They are listed in the comments.
